packages/dds/sequence/src/revertibles.ts

```diff
@@ -190,26 +318,90 @@ function revertLocalPropertyChanged(
 	string.getIntervalCollection(label).changeProperties(id, newProps);
 }
 
-export function revertIntervalRevertibles(string: SharedString, revertibles: IntervalRevertible[]) {
+function newEndpointPosition(
+	offset: number | undefined,
+	restoredSegments: ISegment[],
+	sharedString: SharedString,
+) {
+	if (offset === undefined) {
+		return undefined;
+	}
+
+	let offsetFromSegment = offset;
+	for (const segment of restoredSegments) {
+		if (segment.cachedLength > offsetFromSegment) {
+			return sharedString.getPosition(segment) + offsetFromSegment;
+		}
+		offsetFromSegment -= segment.cachedLength;
+	}
+
+	return undefined;
+}
+
+function revertLocalSequenceRemove(
+	sharedString: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.POSITIONREMOVE>,
+) {
+	const restoredSegments: ISegment[] = [];
+	const saveSegments = (event: SequenceDeltaEvent) => {
+		event.deltaArgs.deltaSegments.forEach((d) => restoredSegments.push(d.segment));
```

Comment:
the ranges on event will already be sorted, and map 1 to 1 with segments

Reply:
could you accelerate insert by building a structure to keep track of the range offset of each range as you grab them to track them?

Reply:
there are a few structures in merge tree that might be able to be applied, red black tree, or SortedSet

Reply:
The ranges in each event will be sorted, but there's no guarantee about the order for separate events? They currently seem to be in order, but I don't think I should rely on that.

Reply:
there are no guarantees across events as events map to ops, and ops are random access.